src/sherpa_ai/output_parsers/entity_validation.py

```diff
@@ -0,0 +1,77 @@
+from typing import Tuple
+
+from sherpa_ai.events import EventType
+from sherpa_ai.memory import Belief
+from sherpa_ai.output_parsers.base import BaseOutputProcessor
+from sherpa_ai.output_parsers.validation_result import ValidationResult
+from sherpa_ai.utils import TextSimilarityState, check_entities_match
+
+
+class EntityValidation(BaseOutputProcessor):
+    """
+     Process and validate the presence of entities in the generated text.
+
+    This class inherits from the BaseOutputProcessor and provides a method to process
+    the generated text and validate the presence of entities based on a specified source.
+
+    Methods:
+    - process_output(text: str, belief: Belief) -> ValidationResult:
+        Process the generated text and validate the presence of entities.
+
+    - get_failure_message() -> str:
+        Returns a failure message to be displayed when the validation fails.
+
+    """
+
+    def process_output(
+        self, text: str, belief: Belief, iteration_count: int = 1
+    ) -> ValidationResult:
+        """
+        Verifies that entities within `text` exist in the `belief` source text.
+        Args:
+            text: The text to be processed
+            belief: The belief object of the agent that generated the output
+            iteration_count (int, optional): The iteration count for validation processing. Defaults to 1.
+        Returns:
+            ValidationResult: The result of the validation. If any entity in the
+            text to be processed doesn't exist in the source text,
+            validation is invalid and contains a feedback string.
+            Otherwise, validation is valid.
+        """
+        source = belief.get_histories_excluding_types(
+            exclude_types=[EventType.feedback, EventType.result],
+        )
+        entity_exist_in_source, error_message = check_entities_match(
+            text, source, self.similarity_picker(iteration_count)
```

Comment:
It seems we are using different similarity-matching approaches at different iterations. Please document this.